packages/contracts-bedrock/src/L2/RevenueSharer.sol

```diff
@@ -0,0 +1,161 @@
+        L1_WALLET = _l1Wallet;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            External Functions
+    //////////////////////////////////////////////////////////////*/
+    /**
+     * @dev Withdraws funds from FeeVaults, sends Optimism their revenue share, and withdraws remaining funds to L1.
+     */
+    function execute() external virtual {
+        // Pull in revenue
+        uint256 d = feeVaultWithdrawal(Predeploys.L1_FEE_VAULT);
+        uint256 b = feeVaultWithdrawal(Predeploys.BASE_FEE_VAULT);
+        uint256 q = feeVaultWithdrawal(Predeploys.SEQUENCER_FEE_WALLET);
+
+        // Compute expenditure
+        uint256 e = getL1FeeExpenditure();
+
+        // Compute revenue and profit
+        uint256 r = d + b + q; // revenue
+        uint256 p = r - e; // profit
+
+        // Compute revenue share
+        uint256 s = Math.max(
+            REVENUE_COEFFICIENT_BASIS_POINTS * r / BASIS_POINT_SCALE,
+            PROFIT_COEFFICIENT_BASIS_POINTS * p / BASIS_POINT_SCALE
+        ); // share
+        uint256 remainder = r - s;
+
+        // Send Beneficiary their revenue share on L2
+        require(SafeCall.send(BENEFICIARY, gasleft(), s), "RevenueSharer: Failed to send funds to Beneficiary");
+
+        // Send remaining funds to L1 wallet on L1
+        L2StandardBridge(payable(Predeploys.L2_STANDARD_BRIDGE)).bridgeETHTo{ value: remainder }(
+            L1_WALLET, WITHDRAWAL_MIN_GAS, bytes("")
+        );
+
+        emit FeesDisbursed(s, r);
+    }
+
+    /**
+     * @dev Returns the RevenueSharer's best estimate of L1 Fee expenditure for the current accounting period.
+     * @dev TODO this just returns zero for now, until L1 Fee Expenditure can be tracked on L2.
+     */
+    function getL1FeeExpenditure() public pure returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @dev Receives ETH fees withdrawn from L2 FeeVaults.
+     * @dev Will revert if ETH is not sent from L2 FeeVaults.
+     */
+    receive() external payable virtual {
+        if (
+            msg.sender != Predeploys.SEQUENCER_FEE_WALLET && msg.sender != Predeploys.BASE_FEE_VAULT
+                && msg.sender != Predeploys.L1_FEE_VAULT
+        ) {
+            revert("RevenueSharer: Only FeeVaults can send ETH to FeeDisburser");
+        }
+        emit FeesReceived(msg.sender, msg.value);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            Internal Functions
+    //////////////////////////////////////////////////////////////*/
+    /**
+     * @dev Withdraws fees from a FeeVault and returns the amount withdrawn.
+     * @param _feeVault The address of the FeeVault to withdraw from.
+     * @dev Withdrawal will only occur if the given FeeVault's balance is greater than or equal to
+     *        the minimum withdrawal amount.
+     */
+    function feeVaultWithdrawal(address payable _feeVault) internal returns (uint256) {
+        require(
```

Comment:
It could be worth also calling to check that the withdrawal network is L2, otherwise the state machine can be fooled